### PR TITLE
add ip limit

### PR DESCRIPTION
**Pull request by ToWorld**
#64

### CODE REVIEW

File: p2pv2/stream_limit.go [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+	if lg == nil {
+		lg = log.New("module", "p2pv2")
+		lg.SetHandler(log.StreamHandler(os.Stderr, log.LogfmtFormat()))
+	}
+	sl.log = lg
+}
+
+func (sl *StreamLimit) hasStream(addrStr string) bool {
+	if sl == nil {
```
Comment (qizheng09) [inline p2pv2/stream_limit.go]: There's no need to check whether `sl` is nil.

File: p2pv2/stream_limit.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+		return false
+	}
+	_, exist := sl.streams.Get(addrStr)
+	return exist
+}
+
+// AddStream used to add the amount of same ip, plus one per call
+func (sl *StreamLimit) AddStream(addrStr string, peerID peer.ID) bool {
+	if sl == nil {
```
Comment (qizheng09) [inline p2pv2/stream_limit.go]: same as above

**Comment by qizheng09** [comment]
LGTM

File: p2pv2/stream_limit.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	"github.com/patrickmn/go-cache"
+
+	"github.com/xuperchain/log15"
+)
+
+// StreamLimit limit the peerID amount of same ip
+type StreamLimit struct {
+	// Store all streams available
+	// key: addr value: peerID
+	streams *cache.Cache
```
Comment (fxsjy) [inline p2pv2/stream_limit.go]: why use cache, not map ?

Comment (ToWorld) [reply p2pv2/stream_limit.go]: You can use map, sync.Map or cache.Cache, No much difference.

Comment (ToWorld) [reply p2pv2/stream_limit.go]: > why use cache, not map ?

I have corrected the issue using sync.Map instead of cache.Cache.

**Pull request merged**